### DISCUSSION TITLE
I'm looking for a better model hook pattern

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1292] [Q&A]
@ganigeorgiev This is how I'm setting up my app after the 0.10 update.  Do you know of any better way to setup model.Record hooks than to create a switch statement on the collection name?

I'm thinking that as it is now that it is the most flexible for any kind of setup, but also it's lacking elegance.  I think that being able to register events for named collections is something so typical that is should be at least considered.  What about changing the hook registration to first have a parameter for the collection name?  For a universal hook it could be either empty string or asterisk.  I think that it's better to control record hooks by registration instead of a static checks in the hook function itself.

```golang
func New() *pocketbase.PocketBase {
	app := pocketbase.New()

	app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		env := env.New(app)

		initCommands(app, env)
		initModelEvents(app, env)

		return nil
	})

	return app
}

func initCommands(app *pocketbase.PocketBase, env *env.Env) {
	rootCmd := app.RootCmd
	rootCmd.AddCommand(cmd.NewWebserverCommand(app, env))
	rootCmd.AddCommand(cmd.NewInstanceCommand(env))
	rootCmd.AddCommand(cmd.NewInstallCommand(env))
	rootCmd.AddCommand(pb_cmd.NewTempUpgradeCommand(app))
}

func initModelEvents(app core.App, env *env.Env) {
	app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		case models.SitesCollectionName:
			return hooks.SiteBeforeCreate(m, env)
		}

		return nil
	})

	app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		case models.SitesCollectionName:
			return hooks.SiteAfterCreate(m, env)
		}

		return nil
	})

	app.OnModelBeforeUpdate().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		}

		return nil
	})

	app.OnModelAfterUpdate().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		}

		return nil
	})

	app.OnModelBeforeDelete().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		}

		return nil
	})

	app.OnModelAfterDelete().Add(func(e *core.ModelEvent) error {
		m := e.Model.(*pbmodels.Record)

		switch m.TableName() {
		}

		return nil
	})
}
```

**Comment by ganigeorgiev** [comment]
@dschissler I think is fine. Some nitpicks:

- The type assertion (`*models.Record`) is not necessary if you want to only access the base `models.Model` interface methods (in your case that is `m.TableName()`).

- You can avoid the central `switch` statement by just calling `app.OnModel*().Add(...)` individually for each of your collections (with the catch that inside the hook you'll have to check the collection name and early return if it is not the one you expect).

---

> What about changing the hook registration to first have a parameter for the collection name?

I think it is doable (at least for the `OnModel*` and `OnRecord*` hooks). We can make the hook getters variadic and accept optional number of table names. It will be mostly a non-breaking change (_unless someone is using directly the `hook.Hook` type result of the getters but I doubt that there are too many users doing it_). Or in other words, it should be possible to do something like:

```go
app.OnModelBeforeCreate("table1", "table2").Add(callback) // the callback will be triggered only on table1 or table2 model create
```

I've added it to the roadmap and eventually will be implemented together with https://github.com/pocketbase/pocketbase/issues/1240.

**Reply by dschissler** [reply]
@ganigeorgiev 

> The type assertion (*models.Record) is not necessary if you want to only access the base models.Model interface methods (in your case that is m.TableName()).

Ok thanks.  I'm at least far enough along in working with the system to know that.  It was a remnant of past experiments.

> You can avoid the central switch statement by just calling app.OnModel*().Add(...) individually for each of your collections (with the catch that inside the hook you'll have to check the collection name and early return if it is not the one you expect).

I don't like this at all, but thanks for offering.

> app.OnModelBeforeCreate("table1", "table2").Add(callback) // the callback will be triggered only on table1 or table2 model create

That is a very good idea to use a variadic parameter.  It doesn't change how it currently is and it offers more.  I think that this will be a huge improvement.  I'll just keep doing what I'm doing for now until this is ready.

**Reply by dschissler** [reply]
@ganigeorgiev Actually, it needs to be cast to Record, because that is what it is for the hook.  I'm not using custom models anymore.  I get what you're saying, but I don't want to have generic hooks at this time that accept a custom model.  It's all pb models.Record from this point out.  It will always be a record.  So I might as well just cast it once.

I think though that, since later each hook will have its own entry point that I'll need to do this casting in each one.

**Reply by ganigeorgiev** [reply]
@dschissler That's ok, just note that the `OnModel*` hooks fire for every `models.Model` and the model type may not be `models.Record` but for example `models.Admin` and the hook could panic. You can add an extra check to ensure that the type assertion is ok, eg. `m, ok := e.Model.(*pbmodels.Record)` (or do the type assertion after `TableName()` check).

**Reply by dschissler** [reply]
@ganigeorgiev Ahh, thanks for that.

Is there any way that we could get a tighter and less problematic event for the boring and typical case of just using Records?  I mean something like a concept that Record is the default or standard way to do things, but that Pocketbase allows a deeper more generic access to for people who want to use custom models (that aren't record inheritors).

**Reply by dschissler** [reply]
@ganigeorgiev I'm thinking like `OnRecord*` events.  My thinking is that it's just much too much involved simply to get a record hook going.  Too many pitfalls and too much knowledge is required to do something that a new user should be able to do very quickly and easily.

**Reply by ganigeorgiev** [reply]
> Is there any way that we could get a tighter and less problematic event for the boring and typical case of just using Records? 

I'm not sure if it is necessary and I'm afraid it may cause some confusion (especially with the `*Request` hooks). Although it is a little verbose at the moment, the type assertion approach is fine and easy to understand.

**Reply by dschissler** [reply]
Maybe this is about nothing because later when you have the variadic addition then won't that be enough to filter it down?  Then it could use the cast without the `ok`.

**Reply by dschissler** [reply]
hey, just an idea and probably not a good one.  Is reflection possible to receive either a model event or a record event?

**Reply by ganigeorgiev** [reply]
I don't understand what do you mean. You can use reflection, if that's what you want, but I don't think it is necessary in this case. `models.Record` implements the `models.Model` interface, aka. every record (and all other internal db mapped structs) are also `models.Model`.

**Reply by dschissler** [reply]
I mean, if Pocketbase used reflection on the model type when calling the hook.  Also with generics.  Anyways, just an idea.

About needing to cast using the form `a, ok := `, I'm thinking that it won't be necessary (for me at least) when you add the variadic event format.  The reason is that I'll only be adding a single collection each time.  So say that it's "articles".  Then on my system it will already be a Record, and it can't be Admin model in my event because I already filtered by the collection when adding the event.

**Comment by dschissler** [comment]
@ganigeorgiev How I'm doing it now

```golang
	app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
		m, ok := e.Model.(*pbmodels.Record)
		if !ok {
			return nil
		}

		switch m.TableName() {
		case models.SitesCollectionName:
			return hooks.SiteBeforeCreate(m, env)
		}

		return nil
	})
```

```golang
// hooks file

func SiteBeforeCreate(m *pbmodels.Record, env *env.Env) error {
	// Do stuff
	return nil
}
```

How it will be after the variadic parameter event setup:

```golang
	app.OnModelBeforeCreate(models.SitesCollectionName).Add(hooks.SiteBeforeCreate)
```

```golang
// hooks file

func SiteBeforeCreate(m *pbmodels.Record, env *env.Env) error {
	// In my situation I don't need to check if it's a Record, because I'm only using records.
	// I also don't need to check if it's a Admin model because I already filtered it by collection when
	// registering the event hook.
	m := e.Model.(*pbmodels.Record)

	// Do stuff
	return nil
}
```

[edit] Sorry, I'll need to wrap `hooks.SiteBeforeCreate` to be able to pass in my custom env.Env structure, but that doesn't change anything.

**Reply by ganigeorgiev** [reply]
Not exactly. The `SiteBeforeCreate` will have to change or you'll have to wrap it because the `Add` func has different definition, aka. it could look something like this:

```go
app.OnModelBeforeCreate(models.SitesCollectionName).Add(func(e *core.ModelEvent) error {
	return yourhooks.SiteBeforeCreate(e.Model.(*pbmodels.Record), env)
})

// yourhooks
func SiteBeforeCreate(record *pbmodels.Record, env *env.Env) error {
	log.Println(record)

	return nil
}
```

In any case, you can also avoid the current type assertion check by doing something like:

```go
app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
	switch e.Model.TableName() {
	case models.SitesCollectionName:
		return hooks.SiteBeforeCreate(e.Model.(*pbmodels.Record), env)
	}

	return nil
})
```

**Reply by dschissler** [reply]
I caught that before you posted.  So I win.

> [edit] Sorry, I'll need to wrap hooks.SiteBeforeCreate to be able to pass in my custom env.Env structure, but that doesn't change anything.

I would probably do this in a helper function that takes the function interface instead since it's constantly duplicated, but is a minor detail.

```golang
func(e *core.ModelEvent) error {
	return yourhooks.SiteBeforeCreate(e.Model.(*pbmodels.Record), env)
}
```

Well though, so if you can implement this then it will shape Pocketbase up nicely.
